### PR TITLE
Make username parsing consistent

### PR DESCRIPTION
**Pull request by akalin-keybase**
Make regexp in libkb/checkers.go more straightforward, and
add tests for CheckUsername.

Make markdown parser implement username check correctly, and
add similar tests.

### CODE REVIEW

**Comment by akalin-keybase** [comment]
r @keybase/react-hackers ? See https://github.com/keybase/client/pull/9928 ... if that PR fixed something, I'd like to know what (and add tests). AFAICT, mentions with underscores renders correctly both with the old and new grammar. @MarcoPolo in case he's still around.

**Comment by akalin-keybase** [comment]
hold off on reviewing, @mmaxim gave me a repro case

File: shared/markdown/parser.pegjs [hunk 1]
```diff
@@ -117,14 +117,20 @@ __INLINE_RULE__<StrikeInline, !StrikeMarker>
...
+ = MentionMarker children:([a-zA-Z0-9]+[a-zA-Z0-9_]?)+ MentionMarker service:ValidMentionService & {
+ const mention = flatten(children)[0]
+ return mention.length >= 2 && mention.length <= 16
+} { return {type: 'mention', children: flatten(children), service: service.toLowerCase()} }
 
 // Same as Mention above, but is just returns text
 // Useful if you don't want a mention in certain contexts (like in a code block)
 MentionlessMention
- = MentionMarker children:([a-zA-Z0-9][a-zA-Z0-9_]*) MentionMarker service:ValidMentionService { return prefix('@', children) }
+ = MentionMarker children:([a-zA-Z0-9]+[a-zA-Z0-9_]?)+ MentionMarker service:ValidMentionService & {
```
Comment (akalin-keybase) [inline shared/markdown/parser.pegjs]: Repeats code from above, but a future PR will make this go away anyway

File: go/libkb/checkers.go [hunk 2]
```diff
@@ -11,7 +11,9 @@ import (
 )
 
 var emailRE = regexp.MustCompile(`^\S+@\S+\.\S+$`)
-var usernameRE = regexp.MustCompile(`^([a-zA-Z0-9][a-zA-Z0-9_]?)+$`)
+
+// Disallow multiple underscores in a row.
+var usernameRE = regexp.MustCompile(`^([a-zA-Z0-9]+[a-zA-Z0-9_]?)+$`)
```
Comment (akalin-keybase) [inline go/libkb/checkers.go]: Adding the inner + makes the regexp clearer IMO, and translates directly to the (greedy) grammar syntax

File: shared/markdown/parser.pegjs [hunk 3]
```diff
@@ -117,14 +117,20 @@ __INLINE_RULE__<StrikeInline, !StrikeMarker>
 Strike
  = StrikeMarker !WhiteSpace children:StrikeInline StrikeMarker !(StrikeMarker / NormalChar) { return {type: 'strike', children: flatten(children)} }
 
-// children grammar adapted from username regexp in libkb/checkers.go.
+// children test adapted from CheckUsername in libkb/checkers.go.
 Mention
- = MentionMarker children:([a-zA-Z0-9][a-zA-Z0-9_]*) MentionMarker service:ValidMentionService { return {type: 'mention', children: flatten(children), service: service.toLowerCase()} }
```
Comment (akalin-keybase) [inline shared/markdown/parser.pegjs]: The problem here with the old grammar `([a-zA-Z0-9][a-zA-Z0-9_]?)+` is that this isn't a regexp, but a grammar, and so would parse greedily and not backtrack. e.g. `m_` and `dan_t` are okay because the underscore falls on an odd index, but `ryan_singer` isn't!

**Comment by akalin-keybase** [comment]
Okay, this is ready now. @maxtaco PTAL at the go side, and @keybase/react-hackers for the JS side.

File: shared/markdown/parser.pegjs [hunk 4]
```diff
@@ -117,14 +117,20 @@ __INLINE_RULE__<StrikeInline, !StrikeMarker>
...
+ = MentionMarker children:([a-zA-Z0-9]+[a-zA-Z0-9_]?)+ MentionMarker service:ValidMentionService & {
```
Comment (buoyad) [inline shared/markdown/parser.pegjs]: could `[a-zA-Z0-9_]?` -> `[_]?` here? if I'm understanding how the grammar works, that would match `[a-zA-Z0-9]+` repeatedly and allow a single underscore in between

Comment (akalin-keybase) [reply shared/markdown/parser.pegjs]: Yes! I don't know why I didn't have that. Will change on Monday...

**Comment by akalin-keybase** [comment]
Okay, comment addressed, PTAL.

File: shared/common-adapters/markdown.shared.js [hunk 5]
```diff
@@ -48,6 +48,7 @@ function preprocessMarkdown(markdown: string, meta: ?MarkdownMeta) {
     return markdown
   }
 
+  // TODO: Allow uppercase in mentions, and just normalize.
   return markdown.replace(/\B@([a-z0-9][a-z0-9_]+)/g, (match, matchedGroup) => {
```
Comment (akalin-keybase) [inline shared/common-adapters/markdown.shared.js]: This regexp is inconsistent with the others, but this will go away in another PR.

File: shared/markdown/parser.pegjs [hunk 6]
```diff
@@ -117,14 +117,20 @@ __INLINE_RULE__<StrikeInline, !StrikeMarker>
 Strike
  = StrikeMarker !WhiteSpace children:StrikeInline StrikeMarker !(StrikeMarker / NormalChar) { return {type: 'strike', children: flatten(children)} }
 
-// children grammar adapted from username regexp in libkb/checkers.go.
+// children test adapted from CheckUsername in libkb/checkers.go.
 Mention
- = MentionMarker children:([a-zA-Z0-9][a-zA-Z0-9_]*) MentionMarker service:ValidMentionService { return {type: 'mention', children: flatten(children), service: service.toLowerCase()} }
+ = MentionMarker children:([a-zA-Z0-9]+"_"?)+ MentionMarker service:ValidMentionService & {
```
Comment (buoyad) [inline shared/markdown/parser.pegjs]: Hmm, I'm curious as to why `@a_b_c` doesn't get parsed correctly as a mention? 
screen shot 2018-01-08 at 10 56 12 am
is the rule to disallow more than one underscore, or just consecutive underscores?

Comment (akalin-keybase) [reply shared/markdown/parser.pegjs]: Note that a user has to exist for it to be highlighted. There doesn't seem to be an `a_b_c` username. I'll add that to the parser tests, though.

Comment (akalin-keybase) [reply shared/markdown/parser.pegjs]: Added, PTAL.
